File: app/services/ai_service.py
import json
import logging
import requests
from typing import Any

logger = logging.getLogger(__name__)

class AIService:
    """
    Dịch vụ AI gọi API của Google Gemini.
    Hỗ trợ cơ chế Fallback tự động nếu gặp lỗi:
    1. gemini-3-flash-preview (Mặc định)
    2. gemini-3-pro-preview
    3. gemini-2.5-flash
    """
    MODELS = [
        "gemini-3-flash-preview",
        "gemini-3-pro-preview",
        "gemini-2.5-flash"
    ]

    def __init__(self, api_key: str):
        self.api_key = api_key

    def call_gemini(self, prompt: str, system_instruction: str = "") -> str:
        """
        Gọi API Gemini với cơ chế Fallback.
        Trả về text kết quả hoặc raise Exception nếu tất cả model đều thất bại.
        """
        if not self.api_key:
            raise ValueError("Chưa cấu hình API Key. Vui lòng vào Cài đặt AI để nhập API Key.")

        last_error = None
        for model in self.MODELS:
            try:
                logger.info(f"Đang thử gọi model {model}...")
                return self._call_model(model, prompt, system_instruction)
            except Exception as e:
                logger.warning(f"Model {model} thất bại: {e}")
                last_error = e
        
        raise RuntimeError(f"Tất cả các model đều thất bại. Lỗi cuối: {last_error}")

    def _call_model(self, model: str, prompt: str, system_instruction: str) -> str:
        url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={self.api_key}"
        
        headers = {
            "Content-Type": "application/json"
        }
        
        payload: dict[str, Any] = {
            "contents": [
                {
                    "parts": [{"text": prompt}]
                }
            ]
        }
        
        if system_instruction:
            payload["systemInstruction"] = {
                "parts": [{"text": system_instruction}]
            }

        response = requests.post(url, headers=headers, json=payload, timeout=30)
        
        if response.status_code != 200:
            error_msg = f"HTTP {response.status_code}: {response.text}"
            raise Exception(error_msg)
            
        data = response.json()
        try:
            return data["candidates"][0]["content"]["parts"][0]["text"]
        except (KeyError, IndexError) as e:
            raise Exception(f"Không thể đọc kết quả từ API trả về: {e}, Response: {data}")

File: js/ai_service.js
class AIService {
    constructor() {
        this.models = [
            "gemini-3-flash-preview",
            "gemini-3-pro-preview",
            "gemini-2.5-flash"
        ];
    }

    getApiKey() {
        return localStorage.getItem('gemini_api_key') || '';
    }

    setApiKey(key) {
        localStorage.setItem('gemini_api_key', key);
    }

    async callGemini(prompt, systemInstruction = "") {
        const apiKey = this.getApiKey();
        if (!apiKey) {
            throw new Error("Chưa cấu hình API Key. Vui lòng nhập API Key để sử dụng tính năng này.");
        }

        let lastError = null;

        for (const model of this.models) {
            try {
                console.log(`Đang thử gọi model: ${model}`);
                return await this._executeRequest(model, apiKey, prompt, systemInstruction);
            } catch (error) {
                console.warn(`Model ${model} thất bại: ${error.message}`);
                lastError = error;
                // Nếu lỗi là 400 (Bad Request do sai API format) hoặc 403 (Invalid Key) thì không nên retry các model sau.
                // Ở đây mô phỏng Fallback nên cứ tiếp tục.
            }
        }

        throw new Error(`Tất cả model đều thất bại. Lỗi cuối: ${lastError.message}`);
    }

    async _executeRequest(model, apiKey, prompt, systemInstruction) {
        const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;
        
        const payload = {
            contents: [{ parts: [{ text: prompt }] }]
        };

        if (systemInstruction) {
            payload.systemInstruction = {
                parts: [{ text: systemInstruction }]
            };
        }

        const response = await fetch(url, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`HTTP ${response.status}: ${errorText}`);
        }

        const data = await response.json();
        try {
            return data.candidates[0].content.parts[0].text;
        } catch (e) {
            throw new Error("Không thể đọc phản hồi từ Gemini API.");
        }
    }
}

// Global instance
window.aiService = new AIService();
